rust/otap-dataflow/crates/engine/src/shared/processor.rs

```diff
@@ -0,0 +1,121 @@
+    /// Unlike receivers, processors have known inputs (messages from previous stages), so the pipeline
+    /// engine can control when to call this method and when the processor executes.
+    ///
+    /// This approach allows for greater flexibility and optimization, giving the pipeline engine
+    /// the ability to decide whether to spawn one task per processor or one task for a group of processors.
+    /// The method signature uses `&mut self` rather than `Box<Self>` because the engine only wants to
+    /// temporarily allow mutation of the processor instance, not transfer ownership.
+    ///
+    /// The processor can:
+    /// - Transform the message and return a new message
+    /// - Filter the message by returning None
+    /// - Split the message into multiple messages by returning a vector
+    /// - Handle control messages (e.g., Config, TimerTick, Shutdown)
+    ///
+    /// # Parameters
+    ///
+    /// - `msg`: The message to process, which can be either a data message or a control message
+    /// - `effect_handler`: A handler to perform side effects such as sending messages to the next node.
+    ///    This can be either Send or !Send depending on the processor's Mode type.
+    ///
+    /// # Returns
+    ///
+    /// - `Ok(())`: The processor successfully processed the message
+    /// - `Err(Error)`: The processor encountered an error and could not process the message
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`Error`] if the processor encounters an unrecoverable error.
+    async fn process(
+        &mut self,
+        msg: Message<PData>,
+        effect_handler: &mut EffectHandler<PData>,
+    ) -> Result<(), Error<PData>>;
+}
+
+/// A `Send` implementation of the EffectHandler.
+#[derive(Clone)]
+pub struct EffectHandler<PData> {
+    core: EffectHandlerCore,
+
+    /// A sender used to forward messages from the processor.
+    msg_sender: tokio::sync::mpsc::Sender<PData>,
+}
+
+/// Implementation for the `Send` effect handler.
+impl<PData> EffectHandler<PData> {
+    /// Creates a new shared (Send) `EffectHandler` with the given processor name.
+    #[must_use]
+    pub fn new(name: Cow<'static, str>, msg_sender: tokio::sync::mpsc::Sender<PData>) -> Self {
+        EffectHandler {
+            core: EffectHandlerCore { node_name: name },
+            msg_sender,
+        }
+    }
+
+    /// Returns the name of the processor associated with this handler.
+    #[must_use]
+    pub fn processor_name(&self) -> Cow<'static, str> {
+        self.core.node_name()
+    }
+
+    /// Sends a message to the next node(s) in the pipeline.
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`Error::ChannelSendError`] if the message could not be sent.
+    pub async fn send_message(&self, data: PData) -> Result<(), Error<PData>> {
+        self.msg_sender
+            .send(data)
+            .await
+            .map_err(|e| Error::ChannelSendError(SendError::Closed(e.0)))?;
+        Ok(())
```

Comment:
I'm probably missing something -- is this the same?
```suggestion
            .map_err(|e| Error::ChannelSendError(SendError::Closed(e.0)))
```

Reply:
Good catch. It's now fixed.